File: python/euler003.py
"""
Project Euler #3: Largest prime factor

The prime factors of 13195 are 5,7,13 and 29. so in this case 29

What is the largest prime factor of a given number ?

"""
import math

def solution(n):
    min_p = get_min_prime_divider(n)
    if min_p == n:
        return n
    else:
        n = n // min_p
        return solution(n)


def get_min_prime_divider(n):
    if n == 2 or n % 2 == 0:
        return 2
    if n <= 1:
        return 1
    for i in range(3, int(math.sqrt(n) + 1), 2):
        if n % i == 0:
            return i

    return n


if __name__ == '__main__':
    # prime 4007972729 so biggest prime divide is itself
    print(solution(4007972729))

File: java/src/euler003.java
/**
 * Project Euler #3: Largest prime factor
 * <p>
 * The prime factors of 13195 are 5,7,13 and 29. so in this case 29
 * <p>
 * What is the largest prime factor of a given number ?
 */
public class euler003 {

    private static long solution(long input) {
        while (true) {
            long minPrime = primeDivider(input);
            if (minPrime < input) {
                input = input / minPrime;
            } else {
                return minPrime;
            }
        }
    }

    private static long primeDivider(long n) {
        if (n <= 1) return 1;
        if (n == 2 || n % 2 == 0) return 2;
        for (long i = 3; i < Math.sqrt(n) + 1; i = i + 2) {
            if (n % i == 0) return i;
        }
        return n;
    }

    public static void main(String[] args) {
        long input = 4007972729l;
        long result = solution(input);
        System.out.println(result);
    }
}
